applied-crypto-examples/src/zksnark.rs

```diff
@@ -0,0 +1,144 @@
+//! An example of naive implementations of zksnarks for tutorial purposes
+
+/// Single root of a polynomial
+#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq)]
+pub struct Root {
+    // a in ax+b
+    a: i64,
+    // b in ax+b
+    b: i64,
+}
+
+/// Unencrypted challenge response pair
+#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq)]
+pub struct UnencryptedChallengeResponse {
+    // p(x) = h(x)*t(x)
+    px: i64,
+    // h(x)
+    hx: i64,
+}
+
+impl UnencryptedChallengeResponse {
+    pub fn new(px: i64, hx: i64) -> Self {
+        Self { px, hx }
+    }
+
+    pub fn verify(&self, x: i64, polynomial: &Polynomial) -> bool {
+        self.px == self.hx * polynomial.eval(x)
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Error {
+    /// Proposed roots would result in a polynomial with coefficients in the rational field
+    OutsideIntegerField(i64, i64),
+}
+
+impl Root {
+    pub fn new(a: i64, b: i64) -> Result<Self, Error> {
+        if b % a == 0 {
+            return Ok(Self { a, b });
+        }
+        Err(Error::OutsideIntegerField(a, b))
+    }
+
+    pub fn eval(&self, x: i64) -> i64 {
+        self.a * x + self.b
+    }
+}
+
+/// Polynomial with coefficients restricted to integers within the field of 8-bit signed integers
+#[derive(Clone, Debug)]
+pub struct Polynomial {
+    // Polynomial roots (a, b) such that a*x + b is a factor of the polynomial
+    roots: Vec<Root>,
+    // public roots
+    public_roots: Vec<Root>,
+}
+
+impl Polynomial {
+    /// Create a new polynomial from a list of roots
+    pub fn new(roots: Vec<Root>) -> Self {
+        Self {
+            roots,
+            public_roots: Vec::new(),
+        }
+    }
+
+    pub fn set_public_roots(mut self, num_public: usize) -> Self {
+        self.public_roots = self.roots[0..num_public].to_vec();
+        self
+    }
+
+    /// Get degree of polynomial
+    pub fn degree(&self) -> usize {
+        self.roots.len()
+    }
+
+    /// Create public polynomial from private polynomial
+    pub fn get_public_polynomial(&self) -> Polynomial {
+        Polynomial::new(self.public_roots.clone())
+    }
```

Comment:
There should be a check here to avoid public polynomials from being generated if no roots are specified

Reply:
Good point. Added.